File: backend/schemas/cad_converter.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class ConverterSettingBase(BaseModel):
    converter_name: str = "ODA File Converter"
    converter_exe_path: str
    output_version: str = "ACAD2018"
    output_type: str = "DXF"
    is_enabled: bool = True


class ConverterSettingCreate(ConverterSettingBase):
    pass


class ConverterSettingUpdate(BaseModel):
    converter_name: str | None = None
    converter_exe_path: str | None = None
    output_version: str | None = None
    output_type: str | None = None
    is_enabled: bool | None = None


class ConverterSettingRead(ConverterSettingBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    last_check_status: str | None
    last_check_message: str | None
    created_at: datetime
    updated_at: datetime


class ConverterCheckResult(BaseModel):
    setting_id: int
    status: str
    message: str


class CadConversionRunRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    batch_id: int
    source_file_id: int
    source_format: str
    target_format: str
    source_path: str
    target_path: str | None
    converter_name: str
    converter_exe_path: str
    status: str
    started_at: datetime
    finished_at: datetime | None
    error_code: str | None
    error_message: str | None
    stdout_log: str | None
    stderr_log: str | None
    created_at: datetime


class DwgConvertResult(BaseModel):
    file_id: int
    status: str
    run_id: int | None = None
    converted_file_path: str | None = None
    error_code: str | None = None
    error_message: str | None = None
    warning_code: str | None = None
    warning_message: str | None = None


class BatchDwgConvertResult(BaseModel):
    batch_id: int
    total_count: int
    success_count: int
    failed_count: int
    skipped_count: int = 0
    items: list[DwgConvertResult]

File: frontend/src/api/cadConverter.ts
import { readApiError } from "./errors";

export type ConverterSetting = {
  id: number;
  converter_name: string;
  converter_exe_path: string;
  output_version: string;
  output_type: string;
  is_enabled: boolean;
  last_check_status: string | null;
  last_check_message: string | null;
  created_at: string;
  updated_at: string;
};

export type ConverterSettingPayload = {
  converter_name: string;
  converter_exe_path: string;
  output_version: string;
  output_type: string;
  is_enabled: boolean;
};

export type ConverterCheckResult = {
  setting_id: number;
  status: string;
  message: string;
};

export type DwgConvertResult = {
  file_id: number;
  status: string;
  run_id: number | null;
  converted_file_path: string | null;
  error_code: string | null;
  error_message: string | null;
  warning_code: string | null;
  warning_message: string | null;
};

export type BatchDwgConvertResult = {
  batch_id: number;
  total_count: number;
  success_count: number;
  failed_count: number;
  skipped_count: number;
  items: DwgConvertResult[];
};

export type CadConversionRun = {
  id: number;
  project_id: number;
  batch_id: number;
  source_file_id: number;
  source_format: string;
  target_format: string;
  source_path: string;
  target_path: string | null;
  converter_name: string;
  converter_exe_path: string;
  status: string;
  started_at: string;
  finished_at: string | null;
  error_code: string | null;
  error_message: string | null;
  stdout_log: string | null;
  stderr_log: string | null;
  created_at: string;
};

export async function getConverterSettings(): Promise<ConverterSetting[]> {
  const response = await fetch("/api/cad/converter-settings");
  if (!response.ok) {
    throw await readApiError(response);
  }
  return response.json() as Promise<ConverterSetting[]>;
}

export async function saveConverterSettings(
  payload: ConverterSettingPayload,
  settingId?: number
): Promise<ConverterSetting> {
  const response = await fetch(
    settingId ? `/api/cad/converter-settings/${settingId}` : "/api/cad/converter-settings",
    {
      method: settingId ? "PATCH" : "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload)
    }
  );
  if (!response.ok) {
    throw await readApiError(response);
  }
  return response.json() as Promise<ConverterSetting>;
}

export async function checkConverterSetting(settingId: number): Promise<ConverterCheckResult> {
  const response = await fetch(`/api/cad/converter-settings/${settingId}/check`, {
    method: "POST"
  });
  if (!response.ok) {
    throw await readApiError(response);
  }
  return response.json() as Promise<ConverterCheckResult>;
}

export async function convertDwgFile(fileId: number): Promise<DwgConvertResult> {
  const response = await fetch(`/api/files/${fileId}/convert-dwg-to-dxf`, {
    method: "POST"
  });
  if (!response.ok) {
    throw await readApiError(response);
  }
  return response.json() as Promise<DwgConvertResult>;
}

export async function convertDwgBatch(batchId: number): Promise<BatchDwgConvertResult> {
  const response = await fetch(`/api/imports/${batchId}/convert-dwg-to-dxf`, {
    method: "POST"
  });
  if (!response.ok) {
    throw await readApiError(response);
  }
  return response.json() as Promise<BatchDwgConvertResult>;
}

export async function listProjectConversionRuns(projectId: number): Promise<CadConversionRun[]> {
  const response = await fetch(`/api/projects/${projectId}/cad-conversion-runs`);
  if (!response.ok) {
    throw await readApiError(response);
  }
  return response.json() as Promise<CadConversionRun[]>;
}

export async function listFileConversionRuns(fileId: number): Promise<CadConversionRun[]> {
  const response = await fetch(`/api/files/${fileId}/cad-conversion-runs`);
  if (!response.ok) {
    throw await readApiError(response);
  }
  return response.json() as Promise<CadConversionRun[]>;
}
